Move color palette editor to asset editor

kimprice opened the pull request:
This is the initial move of the color palette editor from experiments to the asset editor. (Refer to https://github.com/microsoft/pxt-arcade/issues/5327 and https://github.com/microsoft/pxt/pull/9000) The functionality is the same, with minor UI changes.

TO DO: 
- update UI according to new designs,
- add functionality to save/delete custom palettes. Currently, you can have one custom palette, but if you select a built-in palette from the dropdown, you will lose the custom palette.
- make updating the project smoother. Right now, after closing the palette editor modal, it flashes to the blocks view and then goes back to the asset editor view.

Try it out here: https://arcade.makecode.com/app/575d5d9de74b299a768c444de8f27ffb9b964d05-fa87e8108b

![image](https://user-images.githubusercontent.com/15070078/206334392-8827cd9f-7d8a-4b71-9280-1d6a486d8600.png)
![image](https://user-images.githubusercontent.com/15070078/206334433-03f7379a-3b1f-4123-afe3-b5f70f83420c.png)
![image](https://user-images.githubusercontent.com/15070078/206334502-fa71f150-06b6-42e4-85c9-195e447bf818.png)

abchatra commented:
Clicking on the color palette in the build uploaded does nothing for me. Does it work for you

kimprice commented:
> Clicking on the color palette in the build uploaded does nothing for me. Does it work for you

Yes, it works for me. Are you getting any console errors?

srietkerk commented:
This is happening for me as well. 

A gif of what's happening: 
![color-palette](https://user-images.githubusercontent.com/49178322/206525063-9d4a4bee-990b-4524-9fb0-ab071f778c03.gif)


Errors that I'm getting: 
<img width="936" alt="palette-errors" src="https://user-images.githubusercontent.com/49178322/206525376-7e76df0f-58a8-4c43-acc7-24229514db7b.png">

jwunderl commented:
Yeah, I'm getting this error during a react crash: 
```
Uncaught TypeError: Cannot read properties of undefined (reading 'slice')
    at PaletteSwatch (PaletteSwatch.js:7:1)
    at renderWithHooks (react-dom.development.js:14985:1)
    at mountIndeterminateComponent (react-dom.development.js:17811:1)
    at beginWork (react-dom.development.js:19049:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
    at invokeGuardedCallback (react-dom.development.js:4056:1)
    at beginWork$1 (react-dom.development.js:23964:1)
    at performUnitOfWork (react-dom.development.js:22776:1)
    at workLoopSync (react-dom.development.js:22707:1)
```

kimprice commented:
Thanks to all who tested the first link. I fixed the bug and updated the link in the description. 

I was grabbing the initial palette info from the pxt.json file which Joey informed me would have an undefined palette for new projects. The link was working for me because the palette was already defined in my project. I updated it to take the palette info from runtime if the palette in the pxt.json file is undefined.

srietkerk commented on react-common/components/palette/ColorPickerField.tsx:
```diff
@@ -42,24 +21,30 @@ export const ColorPickerField = (props: ColorPickerFieldProps) => {
         if (/#[0-9a-fA-F]{6}/.test(newValue)) {
-            onColorChanged(newValue);
+            onColorChanged(newValue.toUpperCase());
         }
```
I experience some weirdness with this functionality on the first color in the list. If I try to enter any text, it doesn't acknowledge it. This works fine with the other colors in the list. 

https://user-images.githubusercontent.com/49178322/206545046-60389988-0a72-463b-9389-e2ad4ab05c97.mp4

srietkerk replied:
I'm also a bit worried about an edge case where users just spam the keyboard to see what happens in the text field and the color that was there is lost. 

However, I also think that if users are wanting to experiment like that, then it's fair for that progress to be lost. This might be something that we just want to see if users get upset about and change the logic later.

Just wanted to comment as something we might want to keep in mind.

kimprice replied:
I'm having trouble reproducing the bug with the first color. @jwunderl was this happening for you as well?

> I'm also a bit worried about an edge case where users just spam the keyboard to see what happens in the text field and the color that was there is lost.

Yea, ideally the text input wouldn't let you write any non-hex characters. I'm not sure why currently the text validation doesn't happen until on blur. But if the value is invalid, it should change back to the previous value.

jwunderl commented on react-common/components/palette/ColorPickerField.tsx:
```diff
@@ -42,24 +21,30 @@ export const ColorPickerField = (props: ColorPickerFieldProps) => {
+            newValue = "#" + newValue;
+        }
+        if (newValue.length > 7) {
+            newValue = newValue.substring(0,7);
```
```suggestion
            newValue = newValue.substring(0, 7);
```

nit

jwunderl commented on react-common/components/palette/PaletteEditor.tsx:
```diff
@@ -45,22 +53,15 @@ export const PaletteEditor = (props: PaletteEditorProps) => {
-        </div>
-    </Modal>
+    return <div className="common-palette-editor">
+        {(currentPalette).colors.slice(1).map((c, i) =>
```
```suggestion
        {currentPalette.colors.slice(1).map((c, i) =>
```

nit

jwunderl commented on webapp/src/components/assetEditor/assetSidebar.tsx:
```diff
@@ -189,35 +201,42 @@ class AssetSidebarImpl extends React.Component<AssetSidebarProps, AssetSidebarSt
-                    onClick={this.showDeleteModal} />}
-            </List>}
+            <List className="asset-editor-sidebar-controls">
+                    { asset && canEdit && <Button
```
```suggestion
                { asset && canEdit && <Button
```
nit: this section is double indented until the asset palette button; unindenting will also remove a lot of the diff which is nice :)

kimprice replied:
nice catch! thanks

kimprice commented on react-common/components/palette/PaletteEditor.tsx:
```diff
@@ -11,10 +11,18 @@ export interface PaletteEditorProps {
     const { palette, onPaletteChanged } = props;
 
-    const [currentPalette, setCurrentPalette] = React.useState<Palette | undefined>(undefined);
+    const [currentPalette, setCurrentPalette] = React.useState<Palette | undefined>(palette);
```
I'm not sure if one way is better than the other. I am pulling in useEffect, so I could do that for useState too. This `React.useState` is just the way it was written originally. @jwunderl?

jwunderl replied:
It really doesn't matter. You *might* lean towards importing the same way as useEffect just for the slightest consistency, but there is zero performance / behavioral difference

kimprice commented on react-common/components/palette/PaletteEditor.tsx:
```diff
@@ -45,22 +53,15 @@ export const PaletteEditor = (props: PaletteEditorProps) => {
+                index={i}
+                color={c}
+                onColorChanged={newColor => updateColor(i + 1, newColor)}
+                onMoveColor={up => moveColor(i + 1, up)}
```
I'm not able to repro this. Is anyone else able to?

srietkerk replied:
Writing here for transparency Not sure how I got this the first time, but I re-visited the link and haven't gotten this issue again.

riknoll commented on webapp/src/components/assetEditor/assetSidebar.tsx:
```diff
@@ -189,35 +201,42 @@ class AssetSidebarImpl extends React.Component<AssetSidebarProps, AssetSidebarSt
                     className="asset-editor-button"
                     onClick={this.copyAssetHandler} />}
-                {canDelete && <Button
-                    label={lf("Delete Asset")}
```
seems like this change is unrelated to this pr

kimprice replied:
Yea, it's slightly unrelated. We felt the delete button needed to be more consistent with the other asset action buttons in order to better distinguish them from the more high-level palette button. Also aesthetically, it looked odd to have a button with no background in the middle of the set of buttons.

riknoll commented on react-common/styles/palette/PaletteSwatch.less:
```diff
@@ -11,6 +11,9 @@
 .common-palette-color {
     width: 1rem;
     height: 1rem;
+    border-color: black;
+    border-width: thin;
+    border-style: solid;
```
```suggestion
    border: 1px solid #000;
```

riknoll commented:
Maybe don't reload if the palette didn't change?

abchatra commented:
🏆
